File: aun/aun-sdk-core/go/discovery.go
package aun

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sort"
	"time"
)

// GatewayDiscovery Gateway 自动发现
// 通过 well-known URL 获取可用的 Gateway 列表并按优先级排序。
// 与 Python SDK discovery.py 对应。
type GatewayDiscovery struct {
	verifySSL bool
}

// NewGatewayDiscovery 创建 Gateway 发现器
func NewGatewayDiscovery(verifySSL bool) *GatewayDiscovery {
	return &GatewayDiscovery{verifySSL: verifySSL}
}

// gatewayEntry well-known 返回的单条 Gateway 信息
type gatewayEntry struct {
	URL      string `json:"url"`
	Priority int    `json:"priority"`
}

// Discover 从 well-known URL 发现 Gateway
// 返回优先级最高的 Gateway WebSocket URL
func (d *GatewayDiscovery) Discover(ctx context.Context, wellKnownURL string, timeout time.Duration) (string, error) {
	if timeout == 0 {
		timeout = 5 * time.Second
	}

	reqCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	// 创建 HTTP 客户端（根据 verifySSL 配置 TLS）
	transport := &http.Transport{}
	if !d.verifySSL {
		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
	}
	client := &http.Client{Transport: transport}

	req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, wellKnownURL, nil)
	if err != nil {
		return "", NewConnectionError(
			fmt.Sprintf("gateway 发现请求创建失败 (%s): %v", wellKnownURL, err),
			WithRetryable(true),
		)
	}

	resp, err := client.Do(req)
	if err != nil {
		return "", NewConnectionError(
			fmt.Sprintf("gateway 发现失败 (%s): %v", wellKnownURL, err),
			WithRetryable(true),
		)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", NewConnectionError(
			fmt.Sprintf("gateway 发现失败 (%s): HTTP %d", wellKnownURL, resp.StatusCode),
			WithRetryable(true),
		)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", NewConnectionError(
			fmt.Sprintf("读取 gateway 发现响应失败 (%s): %v", wellKnownURL, err),
			WithRetryable(true),
		)
	}

	var payload struct {
		Gateways []gatewayEntry `json:"gateways"`
	}
	if err := json.Unmarshal(body, &payload); err != nil {
		return "", NewValidationError(fmt.Sprintf("解析 well-known 响应失败: %v", err))
	}

	if len(payload.Gateways) == 0 {
		return "", NewValidationError("well-known 返回空的 gateways 列表")
	}

	// 按 priority 排序（升序）
	sort.Slice(payload.Gateways, func(i, j int) bool {
		pi := payload.Gateways[i].Priority
		pj := payload.Gateways[j].Priority
		if pi == 0 {
			pi = 999
		}
		if pj == 0 {
			pj = 999
		}
		return pi < pj
	})

	url := payload.Gateways[0].URL
	if url == "" {
		return "", NewValidationError("well-known 缺少 gateway url")
	}

	return url, nil
}

File: aun/aun-sdk-core/ts/src/discovery.ts
/**
 * Gateway 发现
 *
 * 通过 HTTPS GET 请求 well-known URL 获取 Gateway 列表，
 * 按优先级排序后返回最优 Gateway 的 WebSocket URL。
 *
 * 与 Python SDK 的 GatewayDiscovery 完全对齐。
 */

import { ConnectionError, ValidationError } from './errors.js';

export class GatewayDiscovery {
  private _verifySsl: boolean;

  constructor(opts?: { verifySsl?: boolean }) {
    this._verifySsl = opts?.verifySsl ?? true;
  }

  /**
   * 从 well-known URL 发现 Gateway。
   *
   * @param wellKnownUrl - well-known 发现端点（如 https://alice.aid.com/.well-known/aun-gateway）
   * @param timeout - 请求超时（毫秒，默认 5000）
   * @returns Gateway WebSocket URL
   */
  async discover(wellKnownUrl: string, timeout = 5_000): Promise<string> {
    let payload: Record<string, unknown>;
    try {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), timeout);

      // 使用 Node.js 内置 fetch（18+ 支持）
      // 注意：verify_ssl=false 时需要设置 NODE_TLS_REJECT_UNAUTHORIZED=0 环境变量
      // 或通过自定义 Agent。这里仅标记，实际禁用需在应用层处理。
      const response = await fetch(wellKnownUrl, {
        signal: controller.signal,
      });
      clearTimeout(timer);

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      payload = (await response.json()) as Record<string, unknown>;
    } catch (err) {
      throw new ConnectionError(
        `gateway discovery failed for ${wellKnownUrl}: ${err instanceof Error ? err.message : String(err)}`,
        { retryable: true },
      );
    }

    const gateways = payload.gateways;
    if (!Array.isArray(gateways) || gateways.length === 0) {
      throw new ValidationError('well-known returned empty gateways');
    }

    // 按 priority 排序（数值越小优先级越高）
    const sorted = [...gateways].sort(
      (a, b) => (Number(a?.priority ?? 999)) - (Number(b?.priority ?? 999)),
    );

    const url = sorted[0]?.url;
    if (!url || typeof url !== 'string') {
      throw new ValidationError('well-known missing gateway url');
    }
    return url;
  }
}
